### PR TITLE
Fix mining test

### PR DESCRIPTION
**Pull request by noamnelke**

### CODE REVIEW

File: activation/atxdb.go [hunk 1]
```diff
@@ -491,11 +495,32 @@ func (db *ActivationDb) IsIdentityActive(edId string, layer types.LayerID) (bool
...
+	lastAtxTargetEpoch := atx.PubLayerIdx.GetEpoch(uint16(db.LayersPerEpoch)) + 1
+	if lastAtxTargetEpoch < epoch {
+		db.log.Info("IsIdentityActive latest atx is too old expected=%v actual=%v atxid=%v",
+			epoch, lastAtxTargetEpoch, atx.ShortId())
+		return false, nil
+	}
+
+	if lastAtxTargetEpoch > epoch {
+		if len(ids) < 2 {
+			db.log.Error("IsIdentityActive latest atx is too new but no previous atx len(ids)=%v", len(ids))
```
Comment (antonlerner) [inline activation/atxdb.go]: missing a return error here

File: oracle/blockoracle.go [hunk 2]
```diff
@@ -107,7 +107,8 @@ func (bo *MinerBlockOracle) calcEligibilityProofs(epochNumber types.EpochId) err
 		})
 	}
 	bo.proofsEpoch = epochNumber
-	bo.log.Info("miner \"%v…\" is eligible for blocks on %d layers in epoch %d", bo.nodeID.Key[:5], len(bo.eligibilityProofs), epochNumber)
+	bo.log.Info("miner %v is eligible for %d blocks on %d layers in epoch %d",
+		bo.nodeID.Key[:5], numberOfEligibleBlocks, len(bo.eligibilityProofs), epochNumber)
```
Comment (antonlerner) [inline oracle/blockoracle.go:111]: can replace `bo.nodeID.Key[:5]` with `shortId`

File: oracle/blockoracle.go [hunk 3]
```diff
@@ -62,7 +62,7 @@ func (bo *MinerBlockOracle) BlockEligible(layerID types.LayerID) (types.AtxId, [
 			return types.AtxId{common.HexToHash("EEEEEEEE")}, nil, err
 		}
 	}
-	bo.log.Info("miner \"%v…\" found eligible for %d blocks in layer %d", bo.nodeID.Key[:5], len(bo.eligibilityProofs[layerID]), layerID)
+	bo.log.Info("miner %v found eligible for %d blocks in layer %d", bo.nodeID.Key[:5], len(bo.eligibilityProofs[layerID]), layerID)
```
Comment (antonlerner) [inline oracle/blockoracle.go:65]: can replace `bo.nodeID.Key[:5]` with `shortId`

File: activation/atxdb.go [hunk 4]
```diff
@@ -45,6 +46,9 @@ func (db *ActivationDb) ProcessBlockATXs(blk *types.Block) {
 // ProcessAtx validates the active set size declared in the atx, and validates the atx according to atx validation rules
 // it then stores the atx with flag set to validity of the atx
 func (db *ActivationDb) ProcessAtx(atx *types.ActivationTx) {
+	db.processAtxMutex.Lock()
+	defer db.processAtxMutex.Unlock()
```
Comment (gavraz) [inline activation/atxdb.go:50]: Just reminding you again - defer is very slow and should be avoided. Consider taking care of all returns instead.

File: activation/atxdb.go [hunk 5]
```diff
@@ -491,11 +495,32 @@ func (db *ActivationDb) IsIdentityActive(edId string, layer types.LayerID) (bool
...
 		return false, fmt.Errorf("no active IDs for known node")
 	}
+
 	atx, err := db.GetAtx(ids[len(ids)-1])
 	if err != nil {
 		db.log.Error("IsIdentityActive erred while getting atx err=%v", err)
 		return false, nil
 	}
+
+	lastAtxTargetEpoch := atx.PubLayerIdx.GetEpoch(uint16(db.LayersPerEpoch)) + 1
```
Comment (gavraz) [inline activation/atxdb.go:505]: Add comment to explain what we are doing in this flow.
Also, add a note why it is enough to check only last two.

File: oracle/blockoracle.go [hunk 6]
```diff
@@ -107,7 +107,8 @@ func (bo *MinerBlockOracle) calcEligibilityProofs(epochNumber types.EpochId) err
 		})
 	}
 	bo.proofsEpoch = epochNumber
-	bo.log.Info("miner \"%v…\" is eligible for blocks on %d layers in epoch %d", bo.nodeID.Key[:5], len(bo.eligibilityProofs), epochNumber)
+	bo.log.Info("miner %v is eligible for %d blocks on %d layers in epoch %d",
```
Comment (gavraz) [inline oracle/blockoracle.go:110]: Any reason not to prefer %v?
Using %v will not require you to change the print code if you change the param type.

File: tests/analyse.py [hunk 7]
```diff
@@ -40,8 +40,8 @@ def analyze_mining(deployment,last_layer, layers_per_epoch, layer_avg_size, tota
 
     # assert that each node has created layer_avg/number_of_nodes
     for node in blockmap:
-        first_epoch_blocks = sum([len(blockmap[node]["layers"][str(x)]) for x in range(layers_per_epoch) if str(x) in blockmap[node]["layers"]])
+        blocks_in_relevant_layers = sum([len(blockmap[node]["layers"][str(x)]) for x in range(layers_per_epoch, last_layer) if str(x) in blockmap[node]["layers"]])
```
Comment (gavraz) [inline tests/analyse.py:43]: Consider renaming last_layer to num_of_layers to avoid such misuse again.

**Review by gavraz — APPROVED** [review]

File: oracle/blockoracle.go (hunk 2 above)
Comment (noamnelke) [reply oracle/blockoracle.go:111]: there's no such method here.

File: oracle/blockoracle.go (hunk 3 above)
Comment (noamnelke) [reply oracle/blockoracle.go:65]: same as above

File: oracle/blockoracle.go (hunk 6 above)
Comment (noamnelke) [reply oracle/blockoracle.go:110]: no special reason, but also -- why would I change the param type?

File: tests/analyse.py (hunk 7 above)
Comment (noamnelke) [reply tests/analyse.py:43]: not sure I understand... what's the difference?

File: oracle/blockoracle.go (hunk 6 above)
Comment (gavraz) [reply oracle/blockoracle.go:110]: Why would a programmer change his/someone else's code? (;

File: tests/analyse.py (hunk 7 above)
Comment (gavraz) [reply tests/analyse.py:43]: in the test, the last layer we completed consensus on is 8, the number of layers is 9 (0-8) but last_layer=9 which is confusing...

File: activation/atxdb.go (hunk 1 above)
Comment (noamnelke) [reply activation/atxdb.go]: this shouldn't be an error at all. I changed to `log.Info` and added a `return false, nil`.

File: tests/analyse.py (hunk 7 above)
Comment (noamnelke) [reply tests/analyse.py:43]: I see, I didn't see it that way - I saw it as "the last layer that started", which is 9.

I've already merged this, but if you feel it makes it clearer you can change it in one of your PRs. Both seem equally clear to me, so I don't mind either way.

File: oracle/blockoracle.go (hunk 6 above)
Comment (noamnelke) [reply oracle/blockoracle.go:110]: Let me clarify: `%d` works for all number types. This variable is a number and it doesn't make sense to change it to a non-number type. If that would happen, this log would need to change anyway and the IDE would show a warning on this line.

Comment (noamnelke) [reply oracle/blockoracle.go:110]: I have no objection to using `%v`, but the IDE puts `%d` there when I do `option`+`return` and select `Add format string argument` and I see no reason to change that.

**Pull request merged**